### PR TITLE
review on dex-router

### PR DESCRIPTION
**Pull request by blasrodri**

### CODE REVIEW

File: code/parachain/frame/composable-traits/src/dex.rs [hunk 1]
```diff
@@ -141,7 +141,7 @@ pub trait Amm {
 		pool_id: Self::PoolId,
 		in_asset: AssetAmount<Self::AssetId, Self::Balance>,
 		min_receive: AssetAmount<Self::AssetId, Self::Balance>,
-		keep_alive: bool,
+		keep_alive: bool, // blas: what's the meaning of this param?
```
Comment (KaiserKarel) [inline code/parachain/frame/composable-traits/src/dex.rs:144]: It's a substrate idiom with existential deposit. keep_alive == true means that a swap that causes an account to fall under the existential deposit errors.

File: code/parachain/frame/dex-router/src/lib.rs [hunk 2]
```diff
@@ -163,7 +163,8 @@ pub mod pallet {
 			asset_pair: CurrencyPair<T::AssetId>,
 			route: Option<BoundedVec<T::PoolId, T::MaxHopsInRoute>>,
 		) -> DispatchResult {
-			T::UpdateRouteOrigin::ensure_origin(origin)?;
+			T::UpdateRouteOrigin::ensure_origin(origin)?; // blas: why ignore output value?
```
Comment (KaiserKarel) [inline code/parachain/frame/dex-router/src/lib.rs:166]: Since the UpdateRouteOrigin is RootOrCouncil, this is the standard way to make that generic.

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-865bbb5y0 Pallet Review since last audit](https://app.clickup.com/t/865bbb5y0)

File: code/parachain/frame/dex-router/src/lib.rs [hunk 3]
```diff
@@ -483,24 +484,25 @@ pub mod pallet {
...
-					AssetAmount::new(out_asset_id, T::Balance::zero()),
+					AssetAmount::new(*out_asset_id, T::Balance::zero()), /* blas: why balance::zero()? */
 					keep_alive,
 				)?;
-				in_asset_itr = swap_result.value;
+				in_asset_itr = swap_result.value; // blas: what's this?
 			}
 			ensure!(
-				swap_result.value.amount >= min_receive.amount,
+				swap_result.value.amount >= min_receive.amount, /* blas: why the ensure only on
```
Comment (vimukthi-git) [inline code/parachain/frame/dex-router/src/lib.rs:499]: in dex router swap last result is the one user expects to be about min_receive

File: code/parachain/frame/dex-router/src/lib.rs [hunk 4]
```diff
@@ -483,24 +484,25 @@ pub mod pallet {
...
-					.copied()
-					.find(|a| *a != in_asset_itr.asset_id)
+					.find(|a| a != &&in_asset_itr.asset_id)
 					.ok_or(Error::<T>::NoRouteFound)?;
 				swap_result = T::Pablo::do_swap(
 					who,
 					*pool_id,
 					in_asset_itr,
-					AssetAmount::new(out_asset_id, T::Balance::zero()),
+					AssetAmount::new(*out_asset_id, T::Balance::zero()), /* blas: why balance::zero()? */
```
Comment (vimukthi-git) [inline code/parachain/frame/dex-router/src/lib.rs:493]: this is the min receive amount, for dex router to work we need to allow any amount as the constraints are unknown for intermediate swaps as per current impl.

File: code/parachain/frame/dex-router/src/lib.rs [hunk 5]
```diff
@@ -452,25 +453,25 @@ pub mod pallet {
...
 			// with.
-			let route_iter: &mut dyn Iterator<Item = &T::PoolId> = if !reverse {
-				forward_iter = route.iter();
-				&mut forward_iter
-			} else {
-				backward_iter = route.iter().rev();
-				&mut backward_iter
-			};
+			let mut backward_iter;
+
```
Comment (vimukthi-git) [inline code/parachain/frame/dex-router/src/lib.rs:467]: :+1:

**Review by vimukthi-git — COMMENTED** [review]
Thanks. Just FYI dex-router is not part of the upcoming release.

**Pull request closed without merging**